File: Data Structures & Algorithms/three-integer-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;

        sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size(); i++) {
            for (int j = i + 1; j < nums.size(); j++) {
                for (int k = j + 1; k < nums.size(); k++) {

                    if (nums[i] + nums[j] + nums[k] == 0) {

                        vector<int> target =
                            {nums[i], nums[j], nums[k]};

                        if (find(ans.begin(),
                                 ans.end(),
                                 target) == ans.end()) {

                            ans.push_back(target);
                        }
                    }
                }
            }
        }

        return ans;
    }
};

File: Data Structures & Algorithms/first-missing-positive/submission-0.py
class Solution:
    def firstMissingPositive(self, nums: list[int]) -> int:
        n = len(nums)
        
        for i in range(n):
            # Keep swapping until the element at index i is in its correct slot,
            # or it is out of bounds (<= 0 or > n), or it's a duplicate.
            while 1 <= nums[i] <= n and nums[i] != nums[nums[i] - 1]:
                # Swap nums[i] with the element at its target index
                correct_idx = nums[i] - 1
                nums[i], nums[correct_idx] = nums[correct_idx], nums[i]
                
        # Second pass: find the first index where the value doesn't match the index rule
        for i in range(n):
            if nums[i] != i + 1:
                return i + 1
                
        # If all positions are correct, the missing number is n + 1
        return n + 1